eureka-core/src/main/java/com/netflix/eureka/aws/EIPManager.java

```diff
@@ -63,29 +71,109 @@
+                      ApplicationInfoManager applicationInfoManager) {
+        this.serverConfig = serverConfig;
+        this.clientConfig = clientConfig;
+        this.registry = registry;
+        this.applicationInfoManager = applicationInfoManager;
+        try {
+            Monitors.registerObject(this);
+        } catch (Throwable e) {
+            logger.warn("Cannot register the JMX monitor for the InstanceRegistry", e);
+        }
+    }
 
-    public static EIPManager getInstance() {
-        return s_instance;
+    @PostConstruct
+    public void start() throws Exception {
+        handleEIPBinding();
     }
 
-    private EIPManager() {
-        try {
+    @PreDestroy
+    public void shutdown() throws Exception {
+        timer.cancel();
+        for (int i = 0; i < serverConfig.getEIPBindRebindRetries(); i++) {
+            try {
+                unbindEIP();
+                break;
+            } catch (Exception e) {
+                logger.warn("Cannot unbind the EIP from the instance");
+                Thread.sleep(1000);
+            }
+        }
+    }
 
-            Monitors.registerObject(this);
 
-        } catch (Throwable e) {
-            logger.warn(
-                    "Cannot register the JMX monitor for the InstanceRegistry :",
-                    e);
+    /**
+     * Handles EIP binding process in AWS Cloud.
+     *
+     * @throws InterruptedException
+     */
+    private void handleEIPBinding() throws InterruptedException {
+        int retries = serverConfig.getEIPBindRebindRetries();
+        // Bind to EIP if needed
+        for (int i = 0; i < retries; i++) {
+            try {
+                if (isEIPBound()) {
+                    break;
+                } else {
+                    bindEIP();
+                }
+            } catch (Throwable e) {
+                logger.error("Cannot bind to EIP", e);
+                Thread.sleep(EIP_BIND_SLEEP_TIME_MS);
+            }
         }
+        // Schedule a timer which periodically checks for EIP binding.
+        scheduleEIPBindTask();
+    }
+
+    /**
+     * Schedules a EIP binding timer task which constantly polls for EIP in the
+     * same zone and binds it to itself.If the EIP is taken away for some
+     * reason, this task tries to get the EIP back. Hence it is advised to take
+     * one EIP assignment per instance in a zone.
+     *
+     */
+    private void scheduleEIPBindTask() {
+        timer.schedule(new TimerTask() {
+                           @Override
+                           public void run() {
+                               try {
+                                   // If the EIP is not bound, the registry could  be stale
+                                   // First sync up the registry from the neighboring node before
+                                   // trying to bind the EIP
+                                   if (!isEIPBound()) {
+                                       registry.clearRegistry();
```

Comment:
Do we have to do that? If there is false positive, we will wipe out whole registry, and force all clients to do the full registry fetch. They service is up for the whole time so clients may get empty/partial registry content. As we have eventually consistent system, there is no harm in skipping this step.